### PR TITLE
Mistral: Add support for interleaved attention

### PR DESCRIPTION
**Pull request by manueldeprada**
Adds support for interleaved attention masks to the Mistral model.

### CODE REVIEW

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -347,20 +348,26 @@ def forward(
...
-        causal_mask = mask_function(
-            config=self.config,
-            input_embeds=inputs_embeds,
-            attention_mask=attention_mask,
-            cache_position=cache_position,
-            past_key_values=past_key_values,
-            position_ids=position_ids,
-        )
+        # It may already have been prepared by e.g. `generate`
+        if not isinstance(attention_mask, dict) and self.config.layer_types is not None:
```
Comment (manueldeprada) [inline src/transformers/models/mistral/modeling_mistral.py:352]: asking here: we can also drop the `self.config.layer_types is not None` check and create `layer_types` always on config init. Let me know if that is a better option! :)

**Comment by manueldeprada** [comment]
run-slow: mistral

**Comment by manueldeprada** [comment]
run-slow: mistral

**Comment by manueldeprada** [comment]
run-slow: mistral

File: src/transformers/models/minimax/modular_minimax.py [hunk 2]
```diff
@@ -378,7 +378,10 @@ def forward(
 
 
 class MiniMaxAttention(MixtralAttention):
-    pass
+    def __init__(self, config: MiniMaxConfig, layer_idx: int):
+        super().__init__(config, layer_idx)
+        del is_sliding  # noqa: F821
+        self.sliding_window = getattr(config, "sliding_window", None)
```
Comment (Cyrilvallez) [inline src/transformers/models/minimax/modular_minimax.py]: No change should be needed for MiniMax if Mixtral is correctly updated

File: src/transformers/models/mistral/modeling_mistral.py [hunk 3]
```diff
@@ -135,6 +135,13 @@ def __init__(self, config: MistralConfig, layer_idx: int):
         self.k_proj = nn.Linear(config.hidden_size, config.num_key_value_heads * self.head_dim, bias=False)
         self.v_proj = nn.Linear(config.hidden_size, config.num_key_value_heads * self.head_dim, bias=False)
         self.o_proj = nn.Linear(config.num_attention_heads * self.head_dim, config.hidden_size, bias=False)
+        # This check is necessary to support models that inherit via modular (e.g. Mixtral) and do not use layer_types
+        is_sliding = (
+            config.layer_types[layer_idx] == "sliding_attention"
+            if getattr(config, "layer_types", None) is not None
+            else getattr(config, "sliding_window", None) is not None
+        )
+        self.sliding_window = config.sliding_window if is_sliding else None
```
Comment (Cyrilvallez) [inline src/transformers/models/mistral/modeling_mistral.py]:
```suggestion
        self.sliding_window = config.sliding_window if config.layer_types[layer_idx] == "sliding_attention" else None
```

File: src/transformers/models/mistral/modeling_mistral.py [hunk 4]
```diff
@@ -347,20 +357,29 @@ def forward(
...
         )
+        if not mask_already_prepared:
+            attention_mask = create_masks_for_generate(
+                config=self.config,
+                input_embeds=inputs_embeds,
+                attention_mask=attention_mask,
+                cache_position=cache_position,
+                past_key_values=past_key_values,
+                position_ids=position_ids,
+            )
```
Comment (Cyrilvallez) [inline src/transformers/models/mistral/modeling_mistral.py]: We don't use this function here, see qwen2 for how it's done!

File: src/transformers/models/mixtral/modular_mixtral.py [hunk 5]
```diff
@@ -223,7 +223,10 @@ class MixtralRMSNorm(MistralRMSNorm):
 
 
 class MixtralAttention(MistralAttention):
-    pass
+    def __init__(self, config: MixtralConfig, layer_idx: int):
+        super().__init__(config, layer_idx)
+        del is_sliding  # noqa: F821
+        self.sliding_window = getattr(config, "sliding_window", None)
```
Comment (Cyrilvallez) [inline src/transformers/models/mixtral/modular_mixtral.py]:
```suggestion
        super().__init__(config, layer_idx)
        self.sliding_window = getattr(config, "sliding_window", None)
```

File: src/transformers/models/phi4_multimodal/modular_phi4_multimodal.py [hunk 6]
```diff
@@ -278,7 +277,7 @@ def __init__(
         self.nemo_final_size = length
 
 
-class Phi4MultimodalConfig(Phi3Config):
+class Phi4MultimodalConfig(PretrainedConfig):
```
Comment (Cyrilvallez) [inline src/transformers/models/phi4_multimodal/modular_phi4_multimodal.py]: Why is this config changed?

**Review by Cyrilvallez — COMMENTED** [review]
A few comments to make it fit our standard way of doing 🤗

File: src/transformers/models/mistral/modeling_mistral.py (hunk 4 above)
Comment (manueldeprada) [reply src/transformers/models/mistral/modeling_mistral.py]: thanks!

File: src/transformers/models/phi4_multimodal/modular_phi4_multimodal.py (hunk 6 above)
Comment (manueldeprada) [reply src/transformers/models/phi4_multimodal/modular_phi4_multimodal.py]: oopsie from when I was still figuring out `modular`😅

File: src/transformers/models/mistral/modeling_mistral.py [hunk 7]
```diff
@@ -347,23 +349,36 @@ def forward(
...
+            # Prepare mask arguments
+            mask_kwargs = {
+                "config": self.config,
+                "input_embeds": inputs_embeds,
+                "attention_mask": attention_mask,
+                "cache_position": cache_position,
+                "past_key_values": past_key_values,
+                "position_ids": position_ids,
+            }
+            full_mask_already_prepared = isinstance(attention_mask, torch.Tensor) and len(attention_mask.shape) > 2
```
Comment (manueldeprada) [inline src/transformers/models/mistral/modeling_mistral.py:363]: This check is necessary due to `test_modeling_mistral.py::Mask4DTestHard::test_stacked_causal_mask` passing a raw 4d mask. 

Maybe should I change the test to pass a dict instead? It would break BC...

**Comment by manueldeprada** [comment]
run-slow: mistral

File: src/transformers/models/mistral/modeling_mistral.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/models/mistral/modeling_mistral.py:363]: this does not make sense, the model is sliding only

**Review by ArthurZucker — COMMENTED** [review]
Mistral model was never interleaved, so my first review is to go towards no. 
We can maybe define this in the config but hardcode it to be sliding. Our philosophy is to not add abstraction when it is not relevant. Same for mixtral: modeling code should be unchanged

**Comment by manueldeprada** [comment]
> Mistral model was never interleaved, so my first review is to go towards no. 
We can maybe define this in the config but hardcode it to be sliding. Our philosophy is to not add abstraction when it is not relevant.

[mistralai/Ministral-8B-Instruct-2410](https://huggingface.co/mistralai/Ministral-8B-Instruct-2410) is interleaved, as reported by @hmellor, quoting model's readme _"Trained with a 128k context window with interleaved sliding-window attention"_

I had understood that we wanted to support this 😅 But if Ministral isn’t relevant enough to justify the modeling change, sure just close the PR 🤗 

> We can maybe define this in the config but hardcode it to be sliding. 

The PR already does this: defaults are set to sliding or full attention unless the model's config says otherwise:
https://github.com/huggingface/transformers/blob/2166534a2d9660aadeeb67983925aaad1e4f39f7/src/transformers/models/mistral/configuration_mistral.py#L164-L167
(several Mistral models use full attention, [example here](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2))

> Same for mixtral: modeling code should be unchanged.

This PR doesn’t change Mixtral’s modeling, it only moves `Attention`’s `sliding_window` to `__init__` so inheriting models can modify it.

**Comment by ArthurZucker** [comment]
I don't mind the config, but the modeling code is not longer `mistral`, but `qwen2`. We need to check because that architecture already exists -> does not make sense for me to do an exception here ! So as much as possible let's check which model is thus the closest

**Comment by manueldeprada** [comment]
> I don't mind the config, but the modeling code is not longer `mistral`, but `qwen2`. We need to check because that architecture already exists -> does not make sense for me to do an exception here ! So as much as possible let's check which model is thus the closest

🤯 wow, you have every model on your head! I made a quick Ministral modular from Qwen2 and it matches exactly this PR's outputs with just the bias removal:
```
class MinistralAttention(Qwen2Attention):
    def __init__(self, config, layer_idx: int):
        super().__init__(config, layer_idx)
        self.q_proj = nn.Linear(config.hidden_size, config.num_attention_heads * self.head_dim, bias=False)
        self.k_proj = nn.Linear(config.hidden_size, config.num_key_value_heads * self.head_dim, bias=False)
        self.v_proj = nn.Linear(config.hidden_size, config.num_key_value_heads * self.head_dim, bias=False)
```

So I understand now! Whats the preferred approach?:
1. Give up on Ministral and only do the config change for external libraries (and maybe warn that Ministral may be incorrect on long context windows?).
2. Add a new Ministral model with the slim modular from Qwen2.

**Comment by hmellor** [comment]
https://huggingface.co/mistralai/Ministral-8B-Instruct-2410/blob/main/params.json (Mistral format config) shows that Ministral was supposed to be interleaved.

From what I understand about the history, when this model was originally contributed to Transformers, interleaved sliding attention was not supported, so the model was capped to use all sliding attention.

**Comment by ArthurZucker** [comment]
2. Add a new Ministral model with the slim modular from Qwen2. would be the best IMO! We can open a PR to support or just do a hardcode case for mistral + interleaved -> load Ministral

**Comment by manueldeprada** [comment]
> 2. Add a new Ministral model with the slim modular from Qwen2. would be the best IMO! We can open a PR to support or just do a hardcode case for mistral + interleaved -> load Ministral

I am interested in learning about model bring-up, so I’ll open the PR! Closing this
